src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -630,6 +629,19 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
                     return;
                 }
 
+                if (InitialExecutionContext is null)
```

Comment:
Feels like this should be inside of `BeginRequestProcessing()`

Reply:
I put this here because I figure why bother restoring the ExecutionContext until we're sure there's going to be another request in this loop?

Reply:
I suppose for Http1Connections restoring sooner could unroot stuff before the connection goes idle between requests. I'll move it.